Add map_index to XCom's primary key and interface

uranusjr opened the pull request:
Since the latest XCom migration has not been publicly released, this
modifies that migration in place so we don't need to recreate the XCom
primary key *again*.

While I added the map_index argument to the functions, I'm honestly not
sure how it can be used for most of the interfaces. Those can hopefully
be added incrementally later.

uranusjr committed:
Add map_index to XCom's primary key and interface

Since the latest XCom migration has not been publicly released, this
modifies that migration in place so we don't need to recreate the XCom
primary key *again*.

While I added the map_index argument to the functions, I'm honestly not
sure how it can be used for most of the interfaces. Those can hopefully
be added incrementally later.

ashb commented on airflow/models/xcom.py:
```diff
@@ -108,6 +111,7 @@ def set(
         dag_id: str,
         task_id: str,
         run_id: str,
+        map_index: int = -1,
```
I wonder if instead of all these columns we should take a `Union[TaskInstance, TaskInstanceKey]`?

uranusjr replied:
I was wondering about just `TaskInstanceKey`; it’s quite trivial to build a key from a ti anyway.

ashb replied:
It's `ti.key` infact :)

ashb replied:
Hmmm, I've started making this change and I'm not too sure about it. Will continue for now.

ashb committed:
Change XCom functions to take TIKey instead of n parameters

It's getting to the point where set/clear etc are taking entirely too
many parameters and it's getting unwieldy, and we've already got a class
that encapsulates this info.
